kdc/kerberos5.c

```diff
@@ -38,6 +38,44 @@
 #undef __attribute__
 #define __attribute__(X)
 
+struct kdc_pa_auth_state {
+	int auth_status;
+	const char *auth_details;
+	void *free_ptr;
+};
+
+static krb5_error_code
+_kdc_audit_auth_status(astgs_request_t r,
+		       int auth_status,
+		       const char *auth_details,
+		       const char *pa_type)
+{
+    krb5_context context = r->context;
+    krb5_kdc_configuration *config = r->config;
+    struct HDB *hdb = NULL;
+    int ret = 0;
+
+    if (r->clientdb != NULL) {
+	hdb = r->clientdb;
+    } else if (config->db[0]) {
+	hdb = config->db[0];
+    }
+
+    if (hdb != NULL && hdb->hdb_auth_status != NULL) {
```

Comment:
This is fine but, generally idiomatic to Heimdal is avoiding explicitly specifying NULL